gateway/src/resty/http/proxy.lua

```diff
@@ -133,15 +31,94 @@ end
+      scheme = scheme,
+      host = host,
+      port = port
+    }
+    if scheme == 'https' then
+        options.ssl_server_name = host
+        options.ssl_verify = ssl_verify
+    end
 
+    -- Connect via proxy
     if proxy_uri then
-        return connect_proxy(httpc, request, skip_https_connect)
+        if proxy_uri.scheme ~= 'http' then
+            return nil, 'proxy connection supports only http'
+        else
+            proxy_uri.port = default_port(proxy_uri)
+        end
+
+        -- Resolve the proxy IP/Port
+        local proxy_host, proxy_port = httpc:resolve(proxy_uri.host, proxy_uri.port)
+        local proxy_url = format("%s://%s:%s", proxy_uri.scheme, proxy_host, proxy_port)
+        local proxy_auth = request.proxy_auth
+
+        if scheme == 'http' then
+            -- http proxy needs absolute URL as the request path, lua-resty-http 1.17.1 will
+            -- construct a path_prefix based on host and port so we only set request path here
+            --
+            -- https://github.com/ledgetech/lua-resty-http/blob/master/lib/resty/http_connect.lua#L99
+            request.path = uri.path or '/'
+            options.proxy_opts = {
+                http_proxy = proxy_url,
+                http_proxy_authorization = proxy_auth
+            }
+        elseif scheme == 'https' and skip_https_connect then
+            options.scheme = proxy_uri.scheme
+            options.host = proxy_uri.host
+            options.port = proxy_uri.port
+            options.pool = format('%s:%s:%s:%s', proxy_uri.host, proxy_uri.port, host, port)
+            request.path = format('%s://%s:%s%s', scheme, host, port, request.path or '/')
+
+            local ok, err = httpc:connect(options)
+            if not ok then return nil, err end
+
+            ngx.log(ngx.DEBUG, 'connection to ', proxy_uri.host, ':', proxy_uri.port, ' established',
+                ', pool: ', httpc.pool, ' reused times: ', httpc:get_reused_times())
+
+            ngx.log(ngx.DEBUG, 'targeting server ', host, ':', port)
+
+            local ok, err = httpc:ssl_handshake(nil, host, request.ssl_verify)
+            if not ok then return nil, err end
+
+            return httpc
+        elseif scheme == 'https' then
+            options.proxy_opts = {
+                https_proxy = proxy_url,
+                https_proxy_authorization = proxy_auth
+            }
+        else
+            return nil, 'invalid scheme'
+        end
+
+        -- TLS tunnel is verified only once, so we need to reuse connections only for the same Host header
+        local ok, err = httpc:connect(options)
+        if not ok then return nil, err end
+
+        ngx.log(ngx.DEBUG, 'connection to ', proxy_uri.host, ':', proxy_uri.port, ' established',
+            ', pool: ', httpc.pool, ' reused times: ', httpc:get_reused_times())
+        ngx.log(ngx.DEBUG, 'targeting server ', host, ':', port)
     else
-        return connect_direct(httpc, request)
+        -- Connect direct
```

Comment:
can we add a comment saying that only connections to 3scale backend is using this path?